### PR TITLE
Initial YARA support v1

### PR DESCRIPTION
**Pull request by restrict**
- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/1828).

This PR adds initial support for YARA (https://virustotal.github.io/yara/). Properly configured [file extraction](https://suricata.readthedocs.io/en/latest/file-extraction/file-extraction.html) is required for this to work.

Features implemented:
- The `yara` detection keyword
- Load pre-compiled YARA rules
- Compile and load plaintext YARA rules
- Negation of YARA rules
- Tune the YARA timeout and fast-scan mode using `suricata.yaml`

YARA is not built in by default. Run `configure` with the `--enable-yara` flag to enable it.

Here is an example YARA rule (`example.yara`):
```
import "pe"
import "time"

rule example
{
    strings:
        $magic = { 4d 5a }
    condition:
        $magic at 0 and (pe.timestamp < time.now())
}
```
A Suricata signature using the above YARA rule and storing the matched file:

`alert any any <> any any (msg:"YARA example"; yarac:example.yarac; filestore; sid:1; rev:1;)`

A signature to detect and store new WannaCry samples:

`alert any any <> any any (msg:"YARA New WannaCry sample"; filemd5:!known-wannacry-samples; yarac:wannacry.yarac; filestore; sid:1; rev:1;)`

YARA rules can be compiled ahead-of-time to reduce time spent on launching Suricata and reloading the ruleset. Note the `yarac` keyword used above. This is an alias for the `yara` keyword. The alias can be used to visually indicate that a rule file is pre-compiled.

Rules can be pre-compiled using the [`yarac(1)`](https://yara.readthedocs.io/en/v3.7.1/commandline.html) tool.

`$ yarac example.yara example.yarac`

Here are some other ways to interface with the keyword:

```
yara:example.yara
yara:!example.yarac
yara:"example.yar"
```
Multiple YARA rules can be stored in the same file. Only one YARA match is required to trigger an alert.

The YARA scanner can be configured with `suricata.yaml`. The `timeout` option dictates how many seconds the scanner is allowed to run for. Setting `timeout` to `0` removes the limit. This option is equivalent to calling [`yara(1)`](https://yara.readthedocs.io/en/v3.7.1/commandline.html) with the `-a` / `--timeout` flag. Enabling `fast-scan` can make the scanning faster by avoiding multiple matches on the same string. The fast-scan option has the same effects as using the `-f` / `--fast-scan` flag with [`yara(1)`](https://yara.readthedocs.io/en/v3.7.1/commandline.html).
```
yara:
  timeout: 1
  fast-scan: yes
```

### CODE REVIEW

File: suricata.yaml.in [hunk 1]
```diff
@@ -1560,6 +1560,15 @@ threading:
 luajit:
   states: 128
 
+# YARA options
+yara:
+  # The maximum duration of a scan in seconds. A value of 0
+  # means no timeout. The default is a 1 second timeout.
+  timeout: 1
```
Comment (jasonish) [inline suricata.yaml.in]: Is this a limitation of Yara? What I'm wondering is if ms resolution makes sense here, like 100ms.

File: src/detect-yara.c [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+    SCEnter();
+    int result = 0;
+    int matches = 0;
+    DetectYaraData *yara = (DetectYaraData *)m;
+
+    const uint8_t *data = NULL;
+    uint32_t data_len = 0;
+    uint64_t offset = 0;
+    uint64_t file_size = FileTrackedSize(file);
+    StreamingBufferGetData(file->sb, &data, &data_len, &offset);
```
Comment (jasonish) [inline src/detect-yara.c]: This could return NULL data, would yr_rules_scan_mem be OK with that?

**Comment by jasonish** [comment]
Any chance you can provide a rule, a yara script and a pcap with a matching file so we can add it to https://github.com/OISF/suricata-verify?

File: src/detect-yara.c [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+        goto error;
+
+    return yara;
+
+error:
+    if (yara != NULL)
+        DetectYaraFree(yara);
+    if (yara->filename != NULL)
+        SCFree(yara->filename);
+    return NULL;
```
Comment (jasonish) [inline src/detect-yara.c]: yara has already been freed here. Also, it looks like DetectYaraFree already free's the filename.

**Comment by victorjulien** [comment]
Please submit a new branch & PR addressing the feedback. Thanks!

**Pull request closed without merging**